### PR TITLE
[isolines] Filter geom3 closed degenerate and too small isolines

### PR DESCRIPTION
**Pull request by pastk**
The topography generator tool part from #5556.

Requires re-generation of `*.isolines` files to take effect.

### CODE REVIEW

File: topography_generator/utils/contours.hpp [hunk 1]
```diff
@@ -90,7 +92,14 @@ void SimplifyContours(int simplificationZoom, Contours<ValueType> & contours)
       std::vector<m2::PointD> contourSimple;
       feature::SimplifyPoints(m2::SquaredDistanceFromSegmentToPoint(),
                               simplificationZoom, contour, contourSimple);
-      contour = std::move(contourSimple);
+      // Discard closed lines which are degenerate (<=3 points) or too small for the requested zoom level.
+      /// @todo it doesn't fix all cases as the simplification algo
+      /// can produce e.g. a 5 points closed but degenerate line ("flat", not a loop anymore).
```
Comment (vng) [inline topography_generator/utils/contours.hpp:97]: nit: Looks like this comment is outdated. We filter here all closed degenerated lines.

File: topography_generator/utils/contours.hpp [hunk 2]
```diff
@@ -90,7 +92,14 @@ void SimplifyContours(int simplificationZoom, Contours<ValueType> & contours)
...
+      if (contourSimple.size() > 1 && contourSimple.front() == contourSimple.back() &&
```
Comment (vng) [inline topography_generator/utils/contours.hpp:98]: Can we put CHECK_GREATER(contourSimple.size(), 1, ()); here?

File: indexer/scales.cpp [hunk 3]
```diff
@@ -69,4 +72,17 @@ namespace scales
     // assume that feature is always visible in upper scale
     return (level == GetUpperScale() || std::max(r.SizeX(), r.SizeY()) > GetEpsilonForLevel(level));
   }
+
+  bool IsGoodOutlineForLevel(int level, Points const & poly)
+  {
+    // Areas and closed lines have the same first and last points.
+    // Hence the minimum number of outline points is 4.
+    if (poly.size() < 4)
+      return false;
```
Comment (vng) [inline indexer/scales.cpp:81]: This logic is valid for areas, but I doubt for lines. So we skip all long A-B-A lines? Better keep A-B or every use-case is unique?

Maybe it is valid for isolines, but let us assume that we have say fence from OSM like A-B-A (mistakenly drawn).

File: topography_generator/utils/contours.hpp (hunk 2 above)
Comment (pastk) [reply topography_generator/utils/contours.hpp:98]: nope, SimplifyPoints() can return an empty contourSimple

Comment (vng) [reply topography_generator/utils/contours.hpp:98]: Hm, SimplifyPoints function always keep at least 2 points if we have valid input (>= 2 points)

File: indexer/scales.cpp (hunk 3 above)
Comment (pastk) [reply indexer/scales.cpp:81]: For non-isolines we can convert A-B-A to A-B somewhere around feature_maker (when source OSM ways are processed).
So at this point (after we start using IsGoodOutlineForLevel() in feature_sorter for non-isolines) there should be no such cases.

File: topography_generator/utils/contours.hpp (hunk 2 above)
Comment (pastk) [reply topography_generator/utils/contours.hpp:98]: So if you want to terminate on invalid input then I'll change to CHECK()

Comment (vng) [reply topography_generator/utils/contours.hpp:98]: At least we ensure that GenerateContours makes a good output.

Comment (pastk) [reply topography_generator/utils/contours.hpp]: Ok, I've changed it

**Review by vng — APPROVED** [review]

File: generator/feature_sorter.cpp [hunk 4]
```diff
@@ -368,10 +356,14 @@ bool GenerateFinalFeatures(feature::GenerateInfo const & info, std::string const
...
 
+      LOG(LINFO, ("Writing features' data to", dataFilePath));
+
       // Update bounds with the limit rect corresponding to region borders.
       // Bounds before update can be too big because of big invisible features like a
       // relation that contains an entire country's border.
-      // Borders file may be unavailable when building test mwms.
+      // TODO: Borders file may be unavailable when building test mwms (why?).
+      // Probably its the reason resulting mwm sizes are unpredictable
+      // when building not the whole planet.
```
Comment (vng) [inline generator/feature_sorter.cpp:366]: Leave this comment:
/// @todo m_targetDir is a final MWM folder like 220702, so there is NO borders folder inside,
/// and the next function call always returns false.

**Pull request merged**